### PR TITLE
convert explicit And/Or/Not into unified Operator type

### PR DESCRIPTION
**Pull request by camdencheek**
This is a first step towards resolving the @rvantonder's feedback [here](https://github.com/sourcegraph/sourcegraph/pull/25006#discussion_r715201052). 

Essentially, it unifies the `And`, `Or`, and `Not` types into a single operator type. This makes it easier to do things like query optimization, modification, etc. at the cost of being very slightly more verbose.

I'm honestly not sure what the next step will be, since I don't think we can unify the `Atom` variants under a single type like we do with our query tree since each variant has different required parameters. @rvantonder I'd be interested in pairing on this if you see an opportunity to go further with this.

### CODE REVIEW

File: internal/gitserver/protocol/search.go [hunk 1]
```diff
@@ -86,39 +86,37 @@ func (d DiffModifiesFile) String() string {
...
+	Operands []SearchQuery
 }
 
-func (o Or) String() string {
-	cs := make([]string, 0, len(o.Children))
-	for _, child := range o.Children {
-		cs = append(cs, child.String())
+func (o Operator) String() string {
+	if o.Kind == Not {
+		return "NOT " + o.Operands[0].String()
```
Comment (rvantonder) [inline internal/gitserver/protocol/search.go:104]: Even if `not` is intended to be unary, I would print out all operands here, just so we're not hiding anything if someone programmatically constructs it.

You can factor out a `print list` function if you think it's worth sharing with the operand printer below

**Review by rvantonder — APPROVED** [review]
> at the cost of being very slightly more verbose.

I mean the diff is net negative and I think this makes the logic more terse/clear no? At least it's scanning that way for me at least... :P The only more verbose part is constructing the nodes, and it's better to have functions to do that anyway.

> I'm honestly not sure what the next step will be, since I don't think we can unify the Atom variants under a single type like we do with our query tree since each variant has different required parameters

I think this is great! 🌟  I just think this sets us up for a better direction later. And optimizing Sourcegraph queries with this target type. No, we can't unify the `atom` nodes, it's fine, I don't think there's any more magic here.

Later we can maybe have some fun adding annotations to nodes, to see which operations take a long time after we execute them (calculate query cost). But this isn't something we need right now.

File: internal/gitserver/protocol/search.go [hunk 2]
```diff
@@ -86,39 +86,37 @@ func (d DiffModifiesFile) String() string {
...
+	Or
+	Not
+)
 
-// Or is a predicate that matches if any of its children predicates match
-type Or struct {
-	Children []SearchQuery
+type Operator struct {
+	Kind     OperatorKind
+	Operands []SearchQuery
```
[the remark this answers is not in the document]
Comment (camdencheek) [reply internal/gitserver/protocol/search.go:99]: Yep 😄 Already working on a renaming PR

File: internal/gitserver/search/search_test.go [hunk 3]
```diff
@@ -146,10 +146,13 @@ func TestSearch(t *testing.T) {
...
-			&protocol.DiffMatches{Expr: "lorem"},
-			&protocol.DiffMatches{Expr: "ipsum"},
-		}}
+		query := &protocol.Operator{
+			Kind: protocol.And,
+			Operands: []protocol.SearchQuery{
+				&protocol.DiffMatches{Expr: "lorem"},
+				&protocol.DiffMatches{Expr: "ipsum"},
+			},
+		}
```
[the remark this answers is not in the document]
Comment (camdencheek) [reply internal/gitserver/search/search_test.go:155]: Ah, good call. These nested structs get a bit awkward

**Pull request merged**